iceoryx_utils/include/iceoryx_utils/concurrent/resizeable_lockfree_queue.hpp

```diff
@@ -0,0 +1,198 @@
+    /// during concurrent operation but will always be at most capacity
+    /// @note threadsafe, lockfree
+    /// uint64_t size() const noexcept;
+    using Base::size;
+
+    /// @brief returns the current capacity of the queue
+    /// @return the current capacity
+    /// @note threadsafe, lockfree
+    uint64_t capacity() const noexcept;
+
+    /// @brief inserts value in FIFO order, always succeeds by removing the oldest value
+    /// when the queue is detected to be full (overflow)
+    /// @param[in] value to be inserted is copied into the queue
+    /// @return removed value if an overflow occured, empty optional otherwise
+    /// @note threadsafe, lockfree
+    iox::cxx::optional<ElementType> push(const ElementType& value) noexcept;
+
+    /// @brief inserts value in FIFO order, always succeeds by removing the oldest value
+    /// when the queue is detected to be full (overflow)
+    /// @param[in] value to be inserted is moved into the queue if possible
+    /// @return removed value if an overflow occured, empty optional otherwise
+    /// @note threadsafe, lockfree
+    iox::cxx::optional<ElementType> push(ElementType&& value) noexcept;
+
+    // multiple overloads to set the capacity
+    // 1) The most general one allows providing a removeHandler to specify remove behavior.
+    // 2) The overload where a container is provided can be used if the removed elements are to be stored.
+    //    The container must satisfy certain requirements, such as providing push_back.
+    // 3) The final overload discards removed elements.
+
+    /// @brief      Set the capacity to some value.
+    /// @param[in]  newCapacity capacity to be set
+    /// @param[in]  removeHandler is a function taking an element which specifies
+    ///             what to do with removed elements should the need for removal arise.
+    /// @return     true if the capacity was successfully set, false otherwise
+    template <typename Function,
+              typename = typename std::enable_if<cxx::is_invocable<Function, ElementType>::value>::type>
+    bool setCapacity(uint64_t newCapacity, Function&& removeHandler) noexcept;
+
+    /// @brief Set the capacity to a new capacity between 0 and MaxCapacity, if the capacity is reduced
+    /// it may be necessary to remove the least recent elements.
+    /// @param[in] newCapacity new capacity to be set, if it is larger than MaxCapacity the call fails
+    /// @param[out] removedElements container were potentially removed elements can be stored.
+    /// @return true setting if the new capacity was successful, false otherwise (newCapacity > MaxCapacity)
+    /// @note threadsafe, lockfree but multiple concurrent calls may have no effect
+    template <typename ContainerType = iox::cxx::vector<ElementType, MaxCapacity>,
+              typename = typename std::enable_if<!cxx::is_invocable<ContainerType, ElementType>::value>::type>
+    bool setCapacity(uint64_t newCapacity, ContainerType& removedElements) noexcept;
```

Comment:
The advantage is that you can reuse a container and do not ave to construct a new one from scratch, which depending on the container may be costly. For example our cxx::list constructor is quite a bit slower than std::list from what I know (due to the initialization work it has to perform). 

I can wrap this one in a version returning a container.
The semantics is to add as many removed elements to the container as fit, if it is already full nothing is added. I would prfer not to clear the container and leave it to the caller to provide a container which can fit enough elements.

The queue is resized in any case, the number of elements removed are not known before the actual resize took place and hence you would always have to assume the worst case. Hence the resize cannot and will not fail, but it may fail to give you all removed elements and this can be provided as information in an error. I will create an alternative method and think about whether this one here is better kept private (to implement the wrapping one).

Reply:
We had a discussion before and when you pack this into a minimalistic example into godbolt you see that this is optimized away with a gcc-5.4 with -O1.

Reply:
If you are interested, here is the godbolt link: https://godbolt.org/z/ev6M49 . It is somehow similar to the problem we discussed but in the case of `findService` where the user has to provide a reference to the `instanceContainer`.

Reply:
I forgot an advantage of the current way: template type inference for the container will work. It will not if it is in the return type somehow, than we need to fix the container to be something specific like vector (being less general) or call it with template arguments, leading to bad usability.

I will read this, but I would imagine the variant here performs better than any method returning a new container. Also note that this variant can in principle build by the user using the even more general one were you have to provide a removeHandler (in fact, it currently is). So it is just syntactic sugar.

I am unsure about this one, I cannot see a clearly superior approach and in the end I am also of the opinion that the user has to know what he is doing, we cannot design interfaces for relatively complex structures/algorithms for novice users (but we should strive for a good compromise of performance and ease of use).

Reply:
As discussed with @elBoberido I will remove this method entirely. The user can recreate the behavior with the removeHandler.